### ISSUE TITLE
Too many conflicts for HashMap<UInt64, Mapped, HashCRC32<UInt64>> 

### ISSUE THREAD

**Issue by Alex-Cheng** [ClickHouse/ClickHouse#48639]
The table engine `Join` adopt `HashMap<UInt64, Mapped, HashCRC32<UInt64>>`  as hashmap data structure. I found that if I insert 1M rows into `Join` table, the number of conflicts is huge! That cause the insertion operation costs a lot of time, even infinite. 

My analysis is as following.
-------------------------------------------
`HashMap<UInt64, Mapped, HashCRC32<UInt64>>` is used by table engine `Join` for key type of UInt64.

I began the observation on insertion of data into the hash table `HashMap<UInt64, Mapped, HashCRC32<UInt64>>`, at the time point `t`. The result is listed as following.
```
round	key		hash		init place		final place
t+1	29550181	973285812	207284		267016
t+2	20939309	3926075829	207285		267017
t+3	35626104	1979918772	207284		267018
t+4	14041764	2047027640	207288		267019
t+5	2828012		2852334009	207289		267020

hash(35626104) = 1979918772,   hex(1979918772) = 760329B4
hash(14041764) = 2047027640,   hex(2047027640) = AA0329B9
hash(2828012) = 2852334009,     hex(2852334009) = 7A0329B8

mask = 0xFFFFF
place = hash & mask 

place(1979918772) = 329B4
place(2047027640) = 329B9
place(2852334009) = 329B8
```

The places are very near and the load rate (count of elements in hashtable / capacity of hashtable) is very high(near to 1)

The number of conflicts for insertion of one element is linear to the amount of data which has been inserted.

**Comment by Alex-Cheng** [comment]
Why not use `DefaultHash`?

**Comment by Alex-Cheng** [comment]
After deeper digging, I have some new learning.

<img width="988" alt="image" src="https://github.com/ClickHouse/ClickHouse/assets/1518453/e3a62862-d4c3-43ca-a57d-4540c8aac8f3">

From left to right, let me name the figures as figure1, figure2, figure3, figure4. 
Figure1 shows the memory at the hashtable buf's address. You can see there are some empty cells at buf's address.

Figure2 shows the memory at the buf_addr + <new place>. You can see the memory at/after <new place> is all empty.

Figure3 shows the memory at {buf_addr + <new place> - <collision count>}, which should be the 'origin' place for the new key. You can see there is not any empty cells at the span of (0, collision count). The collision count is as big as 96073.
Figure4 shows the memory at {buf_addr + 0x1FFFF}, as the buf's space is 0x3FFFF and the max_fill is 0x20000. You can see the area of  {buf_addr + 0x1FFFF} are all empty.

<img width="238" alt="image" src="https://github.com/ClickHouse/ClickHouse/assets/1518453/56b839a8-2f23-40fc-81f9-270ff3bdc643">

In conclusion, the data of `HashMap<UInt64, Mapped, HashCRC32<UInt64>>` is not placed in uniform distribution.  I will go on investigation on it.

**Comment by Alex-Cheng** [comment]
![img_v2_1a1d617c-041b-4263-8871-418446b95ccg](https://github.com/ClickHouse/ClickHouse/assets/1518453/562b18c6-fc7e-4e28-a3d6-0817bd225973)
It is my experiment about collision of hash table with CRC32 hash. You can see the lower 24-bit of hash is concentrical while the higher 8-bit largely varies.

**Comment by Alex-Cheng** [comment]
I have created a PR #51014  to resolve it.

**Comment by alexey-milovidov** [comment]
It's not surprising that CRC32-C has a lot of collisions on a specific dataset.
For example, if you concatenate a string with its CRC and then calculate the CRC of the resulting string, it will be zero.

**Comment by nickitat** [comment]
to not leave it as "won't fix", because I guess we'd like to fix this...
I have a dream for some time already to try to rewrite our hash table a little, basically to incorporate design ideas used currently in many good implementations like absl::HT or F14 - have explicit compact metadata stored separately. use simd to match against several cells at once. for this we would likely need a better hash function anyway, but chances are that better lookups could compensate this overhead. also it should provide good perf with much higher load factors (i.e. group by, hash join, hash dict will consume less memory).
but there are also chances that the idea will fail miserably )

**Comment by alexey-milovidov** [comment]
We tried it a few times, and variations of Swiss tables didn't provide speed up for GROUP BY.

**Comment by Alex-Cheng** [comment]
I still want to notify that, in case you may overlooked, I have a dataset to reproduce the collission-related issue. 

[hash_map_crc32_conflict_test_data.txt](https://github.com/ClickHouse/ClickHouse/files/12700858/hash_map_crc32_conflict_test_data.txt)

Just loading the data in above attached file into the HashTable with CRC32 as hash algorithm, can reproduce the issue. The dataset is not intentionaly created for the issue, instead it was created intentionally by a tester for testing another function and by chance spotted this issue.

![image](https://github.com/ClickHouse/ClickHouse/assets/1518453/5c4b168d-44fd-4135-84ab-7499cbb633d6)

**Comment by alexey-milovidov** [comment]
Yes, some datasets will have collisions naturally. For example, if you append crc32 to a string, the resulting string will have zero crc32. While it is good for hash tables most of the time, we should either randomize it or dynamically switch to another hash function.
